pil-analyzer/src/traits_processor.rs

```diff
@@ -0,0 +1,103 @@
+use core::panic;
+use std::collections::{BTreeMap, HashMap};
+
+use powdr_ast::{
+    analyzed::{Expression, FunctionValueDefinition, PolynomialReference, Reference},
+    parsed::{types::Type, TraitImplementation},
+};
+
+pub struct TraitsProcessor<'a> {
+    trait_impls: &'a HashMap<String, Vec<TraitImplementation<Expression>>>,
+}
+
+impl<'a> TraitsProcessor<'a> {
+    pub fn new(trait_impls: &'a HashMap<String, Vec<TraitImplementation<Expression>>>) -> Self {
+        Self { trait_impls }
+    }
+
+    pub fn traits_resolution(
+        &self,
+        references: &mut [&mut Reference],
+        traits_functions_defs: HashMap<String, FunctionValueDefinition>,
+    ) {
+        for r in references.iter_mut() {
+            let Reference::Poly(PolynomialReference {
+                name,
+                type_args: Some(types),
+                ..
+            }) = r
+            else {
+                continue;
+            };
+
+            let Some((full_name, resolved_impl_pos)) =
+                self.resolve_trait_function((name.to_string(), types), &traits_functions_defs)
+            else {
+                continue;
+            };
+
+            let trait_name = full_name
+                .rsplit_once("::")
+                .map(|(prefix, _)| prefix)
+                .unwrap_or(&full_name)
+                .replace("::", "."); // TODO: Remove replace after we merge #1694
+            update_reference(&full_name, r, &resolved_impl_pos, trait_name.to_string());
+        }
+    }
+
+    fn resolve_trait_function(
+        &self,
+        collected_ref: (String, &Vec<Type>),
+        traits_functions_definition: &HashMap<String, FunctionValueDefinition>,
+    ) -> Option<(String, BTreeMap<Vec<Type>, usize>)> {
+        let mut resolved_impl_pos = BTreeMap::new();
+        let Some(FunctionValueDefinition::TraitFunction(ref trait_decl, ref trait_fn)) =
+            traits_functions_definition.get(&collected_ref.0)
+        else {
+            return None;
+        };
+
+        if let Some(impls) = self.trait_impls.get(&trait_decl.name) {
+            for (index, impl_) in impls.iter().enumerate() {
+                let Some(_) = impl_.function_by_name(&trait_fn.name) else {
+                    panic!(
+                        "Could not find function {} for {}",
+                        trait_fn.name, trait_decl.name
+                    );
+                };
+
+                resolved_impl_pos.insert(collected_ref.1.clone(), index);
```

Comment:
I don't understand how this works. Don't you have to check if colelcted_ref.1 unifies with impl_.type_scheme?